### PR TITLE
Issue #12938 Add $params to View::renderDynamic()

### PR DESCRIPTION
**Pull request by mikehaertl**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | yes
| Tests pass?   |  ?
| Fixed issues  | #12938

### CODE REVIEW

File: framework/CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,7 @@ Yii Framework 2 Change Log
 
 - Enh #11058: Add `$checkAjax` parameter to method `yii\web\Controller::redirect()` which controls redirection in AJAX and PJAX requests (ivanovyordan)
 - Enh #12592: Optimized `yii\filters\AccessController` on processing accessrules (dynasource)
+- Enh #12938: Allow to pass additional parameters to `yii\base\View::renderDynamic()`
```
Comment (dynasource) [inline framework/CHANGELOG.md]: please add your name (mikehaertl)

File: framework/base/View.php [hunk 2]
```diff
@@ -334,11 +334,17 @@ public function renderPhpFile($_file_, $_params_ = [])
      * when some portions of the content (called *dynamic content*) should not be cached.
      * The dynamic content must be returned by some PHP statements.
      * @param string $statements the PHP statements for generating the dynamic content.
+     * @param array $params the parameters (name-value pairs) that will be extracted and made
```
Comment (dynasource) [inline framework/base/View.php]: can you add one of your clear examples in the method doc above?

Comment (dynasource) [inline framework/base/View.php]: shouldnt this be:
* 'and be made'
* 'and be reused'

File: framework/base/View.php [hunk 3]
```diff
@@ -334,11 +334,17 @@ public function renderPhpFile($_file_, $_params_ = [])
...
+     * available in the $statement context. The parameters will be stored in the cache and
+     * reused each time $statement is executed. You should make sure, that they are safely serializable.
```
Comment (dynasource) [inline framework/base/View.php]: they = these

**Review by dynasource — CHANGES_REQUESTED** [review]
good addon to the framework. Don't forget the unit tests of course. We want your solution to be perfect now and in the future.

File: framework/base/View.php [hunk 4]
```diff
@@ -334,11 +334,17 @@ public function renderPhpFile($_file_, $_params_ = [])
...
      * @return string the placeholder of the dynamic content, or the dynamic content if there is no
      * active content cache currently.
      */
-    public function renderDynamic($statements)
+    public function renderDynamic($statements, $params = null)
     {
+        if ($params !== null) {
+            $statements = 'extract(unserialize(\'' . serialize($params) . '\'));'.$statements;
```
Comment (samdark) [inline framework/base/View.php]: Need spaces around `.` when concatenating.

**Review by samdark — APPROVED** [review]
Overall all is good.

File: framework/base/View.php (hunk 2 above)
Comment (mikehaertl) [inline framework/base/View.php]: @dynasource I've copied the sentence from the other `$params`, e.g. from `render()`. There's no `be` there either.

**Comment by samdark** [comment]
How about a unit test so it's set in stone and never broken?

**Comment by mikehaertl** [comment]
Regarding the unit tests: There are no tests for `renderDynamic()` yet, at least I couldn't find any. I'm not a real expert on writing framework tests, so I will not find time to write a new test from scratch. Even more as this is a non-trivial task here that involves a complex setup including cache etc.

**Comment by dynasource** [comment]
no problem. Ill add these later.

**Review by dynasource — APPROVED** [review]

File: framework/base/View.php [hunk 5]
```diff
@@ -332,13 +332,26 @@ public function renderPhpFile($_file_, $_params_ = [])
...
+     * ```
      * @param string $statements the PHP statements for generating the dynamic content.
+     * @param array $params the parameters (name-value pairs) that will be extracted and made
+     * available in the $statement context. The parameters will be stored in the cache and be reused
+     * each time $statement is executed. You should make sure, that these are safely serializable.
      * @return string the placeholder of the dynamic content, or the dynamic content if there is no
      * active content cache currently.
      */
-    public function renderDynamic($statements)
+    public function renderDynamic($statements, $params = null)
```
Comment (rob006) [inline framework/base/View.php]: Why not typehint `$params` as array?

Comment (mikehaertl) [inline framework/base/View.php]: I found it easier to check `$params === null` than `count($params) === 0`. But I don't really mind to change this. We should also use `@param array|null ...` in the comment block if we keep `null`.

Comment (rob006) [inline framework/base/View.php]: With

```php 
public function renderDynamic($statements, array $params = null)
```

you could use `$params !== []` or just `!empty($params)`.

Comment (mikehaertl) [inline framework/base/View.php]: @dynasource Any preferences? Should I change it?

Comment (rob006) [inline framework/base/View.php]: I fixed my example.

Comment (dynasource) [inline framework/base/View.php]: Im working on the tests now (encountering some problems with testing eval()). Will include the array hinting, so you're allowed to take a break ;)

**Review by samdark — APPROVED** [review]

**Review by rob006 — CHANGES_REQUESTED** [review]
This change should be noted in `UPGRADE.md`.

**Comment by dynasource** [comment]
thanks to you all!

File: framework/base/View.php (hunk 5 above)
Comment (samdark) [inline framework/base/View.php]: @mikehaertl `$array !== []` with array type hinting is preferred.

Comment (rob006) [inline framework/base/View.php]: @samdark Why? `empty()` is faster than `=== []` and IMO more readable.

Comment (dynasource) [inline framework/base/View.php]: although the performance differences are minor, empty is indeed more readable

Comment (samdark) [inline framework/base/View.php]: @rob006 is it faster?

Comment (samdark) [inline framework/base/View.php]: Indeed it is. Let's use it then.

Comment (mikehaertl) [inline framework/base/View.php]: @rob006 @samdark Any references? And is this only for empty arrays? For many years now we've seen `===` as the fastest way to compare e.g. for `null` or empty strings. It's still used in many places all over the framework. Did this change?

Comment (samdark) [inline framework/base/View.php]: I've made a test for `empty($array)` and `$array !== []` for both empty and non-empty arrays at https://3v4l.org/.

Comment (dynasource) [inline framework/base/View.php]: yeah, did the same for `isset()` vs `===` a week ago. Therefore I prefer isset() from now. Faster and more readable

Comment (mikehaertl) [inline framework/base/View.php]: @samdark @dynasource Hmm, I'd vote against that. Unless there's a reliable "official" source that `isset` and `empty` are really always faster, why suddenly change what was considered best practice over all the years? It's like a convention in the Yii framework to always use `===` whenever applicable.

I also don't agree that `isset()` is more readable than `===null`. It's not even the same thing. `isset()` also works for undefined variables. Something you don't always want.

So before changing this convention lightheartedly we should really ask, if this is really neccessary. "If it ain't broken - don't fix it!".

Comment (dynasource) [inline framework/base/View.php]: * https://3v4l.org/ can show you the prove. You can try it yourself. @samdark, where your full url :).
* isset() is more readable. You read from left to right. Isset() shows immediately whats going on, compared to `===` which can be a equation to anything
* 'It's not even the same thing': agree, which shows you, you have to use it when it is appropiate

Comment (samdark) [inline framework/base/View.php]: @mikehaertl it's not suddenly. Actually [it's in our code style](https://github.com/yiisoft/yii2/blob/master/docs/internals/core-code-style.md#--vs-empty) for quite a while...

Comment (mikehaertl) [inline framework/base/View.php]: @samdark If it's only about arrays I'm fine. It wasn't me who brought up the `isset()` topic ;).

Still the `===null` vs `isset()` convention should also be part of the coding style guide.

Comment (dynasource) [inline framework/base/View.php]: :)

**Comment by srnden** [comment]
I would like to see this change in 2.0.*

**Comment by samdark** [comment]
It won't be merged since 2.0.x doesn't accept features from last year.

**Pull request merged**